cli/data.go

```diff
@@ -35,89 +37,182 @@ func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
 		return errors.Wrapf(err, "error creating destination directories")
 	}
 
-	var last string
-	numFilesDownloaded := 0
-	for {
-		var err error
-		var resp *datapb.BinaryDataByFilterResponse
-		for count := 0; count < maxRetryCount; count++ {
-			resp, err = c.sendBinaryDataByFilterRequest(filter, last)
-			if err == nil {
+	if concurrentDownloads == 0 {
+		concurrentDownloads = defaultConcurrentDownloads
+	}
+
+	ids := make(chan string, concurrentDownloads)
+	// Give channel buffer of 1+concurrentDownloads because that is the number of goroutines that may be passing an
+	// error into this channel (1 get ids routine + concurrentRequest download routines).
+	errs := make(chan error, 1+concurrentDownloads)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg := sync.WaitGroup{}
+
+	// In one routine, get all IDs matching the filter and pass them into ids.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var limit int
+		if concurrentDownloads > 100 {
```

Comment:
I think we still want to support parallel values higher than this limit, so I'm going to still allow arbitrary values for parallel. This will just get the IDs in multiple requests, but the channel feeding them into the download routine is still parallelRequests big, so we can still do parallelRequests at a time

Added maxLimit const